### PR TITLE
[feat] technology filter functionality

### PR DESCRIPTION
**Pull request by ethan-tam33**
### Description
[//]: # "[Required] Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."

- markers and clusters on the map and projects in the modal appear according to the selected filters in the Technology Dropdown and Search Bar
- searching in project modal searches within the filtered set of projects
- tested and added a `clearFilters` function for future use

### Screenshots
[//]: # "Add any screenshots that will help your PR reviewer understand what’s going on in the PR!"
[//]: # "[Required] For frontend tasks: add screenshots of the layout and expected behavior (GIFs if you’re feeling extra fancy)!"
[//]: # "Use the provided image template. Drag the desired image into the PR, then copy the link into the placeholder."

[image placeholder]: 

https://github.com/user-attachments/assets/7cb98f75-09db-487a-9ed0-5997ffbfafbb

## How to review
[//]: # "/[Required] Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?"

- review `MapViewScreen.tsx` for all filtering logic

## Next steps
[//]: # "[Optional] - What has not yet been added to the PR, doesn’t yet have any functionality — essentially, what needs to be fixed or added."
[//]: # "[Optional] any temporary fixes in this PR that should be cleaned up later."

- use `clearFilters` function in a new button or existing UI component
- add filter functionality for Status filter

### Online sources
[//]: # "[Optional] insert links to any useful tutorials or documentation that you used for this PR!"

### Related PRs
[//]: # "Add related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

CC: @itsliterallymonique

### CODE REVIEW

File: api/maps/AddMarkers.tsx [hunk 1]
```diff
@@ -92,15 +94,19 @@ export default function AddMarker({
...
 
-  const clusterer = useMemo(() => {
-    if (!map) return null;
+  useEffect(() => {
+    if (!map || !projects) return;
+
+    // clear previous cluster markers
+    if (clusterer) {
+      clusterer.clearMarkers();
+    }
```
Comment (itsliterallymonique) [inline api/maps/AddMarkers.tsx:103]: Your sprint instructions did not tell you to implement this functionality. 
This implementation re-renders the map each time, which makes an API call. So if I search 'abcdefg', I will be making 7 API calls which will become costly. 
As per my sprint assignments, Neha will be implementing this functionality in a way that will make less API calls and we will be merging her implementation.

File: components/MapViewScreen/index.tsx [hunk 2]
```diff
@@ -51,20 +51,54 @@ export default function MapViewScreen({
...
   });
+  const [filteredWithoutSearch, setFilteredWithoutSearch] =
+    useState<Project[]>(projects);
 
-  const handleFilterChange = (filter: FilterType) => {
-    console.log(filter);
+  // show projects based on selected filters
+  const handleFilterButtonClick = () => {
+    const { status, technology, projectSize, location } = selectedFilters;
+    console.log(
```
Comment (itsliterallymonique) [inline components/MapViewScreen/index.tsx:60]: Before you submit your sprint for review + merging, you should remove all testing console log statements because we don't want this on the deployed app.

**Review by itsliterallymonique — CHANGES_REQUESTED** [review]
Please make the changes based on the comments I made.

File: components/MapViewScreen/index.tsx [hunk 3]
```diff
@@ -51,20 +51,54 @@ export default function MapViewScreen({
...
+    // add all filtering logic here
+    const filteredProjects = projects?.filter(project =>
+      technology.includes(project.renewable_energy_technology),
+    );
+    setFilteredProjects(filteredProjects);
+    setFilteredWithoutSearch(filteredProjects);
   };
 
+  // clear filters
+  // const clearFilters = () => {
```
[the remark this answers is not in the document]
Comment (ethan-tam33) [reply components/MapViewScreen/index.tsx]: done

File: api/maps/AddMarkers.tsx (hunk 1 above)
Comment (ethan-tam33) [reply api/maps/AddMarkers.tsx:45]: The code here is for updating clusters for filtering, not for re-rendering the map on search (but I fixed that now). If the clusters are not updated, then when filters are applied from the dropdown, the clusters from the markers that are not part of the filters remain on the map.

File: components/MapViewScreen/index.tsx (hunk 2 above)
Comment (ethan-tam33) [reply components/MapViewScreen/index.tsx]: done

File: components/MapViewScreen/index.tsx [hunk 4]
```diff
@@ -52,19 +52,43 @@ export default function MapViewScreen({
...
         project.project_name.toLowerCase().includes(searchTerm.toLowerCase()),
-      ) ?? null;
+      ) ?? [];
+    setFilteredProjectsWithSearch(searchedProjects);
+  }, [projects, searchTerm]);
 
-    setFilteredProjects(filtered);
-  }, [projects, searchTerm, setFilteredProjects]);
+  const handleFilterChange = (filter: FilterType) => {
+    console.log(filter);
```
Comment (itsliterallymonique) [inline components/MapViewScreen/index.tsx:90]: Please delete all console.log statements. As I mentioned before, we don't want them in our deployed app.

File: api/maps/AddMarkers.tsx (hunk 1 above)
Comment (itsliterallymonique) [reply api/maps/AddMarkers.tsx:45]: I commented the exact line I am talking about. You are re-rendering the map when we filter, which we can't do because that is an API call. Please fix this.

Also, the updating clusters for filtering was also not part of your sprint.

File: components/MapViewScreen/index.tsx [hunk 5]
```diff
@@ -74,9 +98,11 @@ export default function MapViewScreen({
         onFilterChange={handleFilterChange}
         selectedFilters={selectedFilters}
         setSelectedFilters={setSelectedFilters}
+        handleFilterButtonClick={handleFilterButtonClick}
+        clearFilters={clearFilters}
       />
-      <Map projects={projects} />
-      <ProjectsListingModal projects={filteredProjects} />
+      <Map projects={filteredProjects} />
```
Comment (itsliterallymonique) [inline components/MapViewScreen/index.tsx:104]: In this line, you are re-rending the map each time by passing in filteredProjects as a prop. Please change it back to projects. As I mentioned before, filtering the markers was Neha's sprint task, and the implementation she has will be making less API calls than your implementation.

**Review by itsliterallymonique — CHANGES_REQUESTED** [review]

File: components/MapViewScreen/index.tsx [hunk 6]
```diff
@@ -51,20 +51,63 @@ export default function MapViewScreen({
...
+    setFilteredProjects(filteredProjectsFromDropdowns);
+  }, [filteredProjectsFromDropdowns, setFilteredProjects]);
+
+  useEffect(() => {
+    setFilteredProjects(filteredProjectsFromSearch);
+  }, [filteredProjectsFromSearch, setFilteredProjects]);
+
+  const handleFilterChange = (filter: FilterType) => {
+    return;
+  };
```
Comment (ronniebeggs) [inline components/MapViewScreen/index.tsx:110]: Is this needed?

File: components/MapViewScreen/index.tsx [hunk 7]
```diff
@@ -51,20 +51,63 @@ export default function MapViewScreen({
...
+    let projectsToSearch: Project[] = [];
+
+    if (filteredProjectsFromDropdowns.length > 0) {
+      projectsToSearch = filteredProjectsFromDropdowns;
+    } else {
+      projectsToSearch = projects;
+    }
+
+    const searchedProjects: Project[] =
+      projectsToSearch?.filter(project =>
```
Comment (ronniebeggs) [inline components/MapViewScreen/index.tsx:93]: Can filters be applied after a search is made? In that case, would it make sense to search then filter?

I wonder if there's a way to treat "search" similar to a filter criteria. Each filter would then be applied sequentially in the order they're selected.

**Review by ronniebeggs — COMMENTED** [review]
I left a few mostly conceptual comments regarding the filter process. Looking good so far! Props for tackling a fairly complex feature! @ethan-tam33 @itsliterallymonique

File: components/TechnologyDropdown/styles.ts [hunk 8]
```diff
@@ -72,6 +72,12 @@ export const ButtonWithIconStyles = styled.div`
   cursor: pointer;
 `;
 
+export const FilterIconStyles = styled.div`
+  display: flex;
+  align-items: center;
+  flex-direction: row;
+`;
```
[the remark this answers is not in the document]
Comment (ethan-tam33) [reply components/TechnologyDropdown/styles.ts:79]: Hmm previously only `ButtonWithIconStyles` had the `onClick` property, but I had to manually add it for each of the sub-components in this sprint because of the temporary `clearFilters` method on the exit button. I don't think this change will be permanent, but I do agree that we should look into making `ButtonWithIconStyles` extend a button instead of a div.

File: components/MapViewScreen/index.tsx [hunk 9]
```diff
@@ -51,20 +51,63 @@ export default function MapViewScreen({
...
+      projectSize: { min: 0, max: 0 },
+      location: [],
+    });
+    setFilteredProjects(projects);
+  };
+
+  // show projects based on selected filters
+  const handleFilterButtonClick = () => {
+    /* eslint-disable @typescript-eslint/no-unused-vars */
+    const { status, technology, projectSize, location } = selectedFilters;
```
[the remark this answers is not in the document]
Comment (ethan-tam33) [reply components/MapViewScreen/index.tsx:74]: Ohh I see, that's a good idea! I made the filters under the impression that they wouldn't need to be changed, but I agree that making the logic more flexible would be better in the long run. 

The way filtering works currently is that I have a pre-defined object of possible filters (`selectedFilters`) that the projects can filter through within the `handleFilterButtonClick` method. `selectedFilters` gets updated every time a new filter is applied or un-applied. Is there anything specifically you would want me clarify?

File: components/MapViewScreen/index.tsx (hunk 6 above)
Comment (ethan-tam33) [reply components/MapViewScreen/index.tsx:110]: ah nope, deleted!

File: components/MapViewScreen/index.tsx (hunk 7 above)
Comment (ethan-tam33) [reply components/MapViewScreen/index.tsx:93]: The way that I've implemented it allows for filtering after a search is made! I think it does make sense to search then filter? For example, if someone wanted to checkout a project that has a common word "north", "south", etc. and then wanted to filter by dropdowns. This is probably a niche, rare use case but the way that the filtering is implemented allows for search then filter and filter then search to produce the same outputs in the project modal.

Do you mind explaining your idea of sequentially applying filters? I'm not quite sure I understand how the projects would be rendered differently.

**Comment by ethan-tam33** [comment]
@ronniebeggs Thanks for the comments! I'll squash and merge once we resolve our conversations here.

File: components/MapViewScreen/index.tsx (hunk 7 above)
Comment (ronniebeggs) [reply components/MapViewScreen/index.tsx:99]: Ok it seems that the implementation already works as intended then. I was just curious about the ordering that filters are applied, because it could affect performance. 

For example, if you wish to search for "north" and filter by "hydroelectric," it might make sense to filter first then search, because depending on the search implementation, I imagine search would be an asymptotically slower operation (??), and thus better to perform on a smaller subset of projects.

The more I think about it, this is entirely dependent on search logic and how project data is stored, which I don't have much context on. Just something to consider!

Comment (ethan-tam33) [reply components/MapViewScreen/index.tsx:99]: Hmm are you both suggesting that we prevent any searching until the user clicks on at least one filter? 

Regarding the time complexity of searching, I believe the way that we have set it up has filtering for search and dropdown filters to have the same complexity. We're iterating over each project and checking if it includes the search term or is included in the dropdown filters.

File: components/MapViewScreen/index.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (ethan-tam33) [reply components/MapViewScreen/index.tsx:80]: Ohh I see, I think I'm doing that? The `technology` variable in line 80 is just a list that holds a bunch of strings (or is empty). Then line 82 filters all the projects by checking if its `renewable_energy_technology` attribute is within this `technology` list, without caring about specific values. Is this what you mean about not hardcoding the attributes?

[the remark this answers is not in the document]
Comment (ronniebeggs) [reply components/MapViewScreen/index.tsx:80]: I guess, given that this PR is only for the technology filter, your implementation suffices. I was just curious if there was a way to iterate over the `selectedFilters` object so we could easily implement `status`, `projectSize`, and `location` filters as well. However doing so in a way where each filter is it's own function/reducer that, when applied to a given row, can immediately determine whether it belongs in the output. Does that make sense? It might not be necessary for this implementation, but it's one possible approach.

File: components/MapViewScreen/index.tsx (hunk 7 above)
Comment (ronniebeggs) [reply components/MapViewScreen/index.tsx:99]: I see that search isn't very complex in this case. Although I think the reason I bring this up is based off the main `useEffect`, where `projectsToSearch` selects all filtered projects, then the search term is applied to this list. Suppose I have filtered by `hydroelectric` then searched for `"north"`. I'd now have a list of projects satisfying these criteria. If I additionally applied the `operational` filter, ideally I'd take the list of already filtered projects, and apply the additional filter there. However, it seems that filters are applied first, separately from search, then the search term is applied after. Do you see what I mean? I'm not sure if that makes sense or whether it's actually needed, but something I noticed.

Comment (ethan-tam33) [reply components/MapViewScreen/index.tsx:99]: Ohh ok I see what you are saying! I agree that we should do this for filters part of the `Status` filter. However, I don't think we can do this for a flow like `hydroelectric` -> search for `north` -> `land-based wind` (aka another `technology` filter). We would need to filter all projects based on the `technology` filters and those wouldn't exist in the modal prior to selecting `land-based wind`. Perhaps this is a task that Neha's recent PR can reflect then?

File: components/MapViewScreen/index.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (ethan-tam33) [reply components/MapViewScreen/index.tsx:80]: Ohh I see now, you're referring to the general filters. I think we can definitely do that and it makes sense to do so- this probably will involve some mapping of the filters to the project attributes in question. And then the method can just do an agnostic loop over that mapping to filter the projects. Thanks Ronnie! I'll adjust my PR for this.

File: components/MapViewScreen/index.tsx (hunk 7 above)
Comment (itsliterallymonique) [reply components/MapViewScreen/index.tsx:99]: For now, just squash and merge your sprint so that I can check everything works smoothly for this week's sprints. We can sort this out another time.
